nni/compression/pytorch/utils/counter.py

```diff
@@ -24,112 +237,102 @@ def count_flops_params(model: nn.Module, input_size, custom_ops=None, verbose=Tr
-    model : nn.Module
+    model: nn.Module
         target model.
-    input_size: list, tuple
-        the input shape of data
+    x: tuple or tensor
+        the input shape of data or a tensor as input data
     custom_ops: dict
-        a mapping of (module: custom operation)
-        the custom operation will overwrite the default operation.
-        for reference, please see ``custom_mask_ops``.
+        a mapping of (module -> torch.nn.Module : custom operation)
+        the custom operation is a callback funtion to calculate
+        the module flops and parameters, it will overwrite the default operation.
+        for reference, please see ``DEFAULT_OPS``.
+    verbose: bool
+        If False, mute detail information about modules. Default is True.
+    mode:
+        the mode of how to collect information. If the mode is set to `default`,
+        only the information of convolution and linear will be collected.
+        If the mode is set to `full`, other operations will also be collected.
 
     Returns
     -------
-    flops: float
+    flops: int
         total flops of the model
-    params:
+    params: int
         total params of the model
+    results: dict
+        the detail information of modules. (module_name, module_type, weight_size,
+        flops, params, input_size, output_size) are included in the results.
+
     """
 
-    assert input_size is not None
+    assert isinstance(x, tuple) or isinstance(x, torch.Tensor)
+    assert mode in ['default', 'full']
 
-    device = next(model.parameters()).device
-    inputs = torch.randn(input_size).to(device)
+    original_device = next(model.parameters()).device
+    training = model.training
 
-    hook_module_list = []
-    if custom_ops is None:
-        custom_ops = {}
-    custom_mask_ops.update(custom_ops)
-    prev_m = None
-    for m in model.modules():
-        weight_mask = None
-        m_type = type(m)
-        if m_type in custom_mask_ops:
-            if isinstance(prev_m, PrunerModuleWrapper):
-                weight_mask = prev_m.weight_mask
+    if torch.is_tensor(x):
+        x = (t.to(original_device) for t in x)
+    else:
+        x = (torch.zeros(x).to(original_device), )
 
-            m.register_buffer('weight_mask', weight_mask)
-            hook_module_list.append(m)
-        prev_m = m
 
-    flops, params = profile(model, inputs=(inputs, ), custom_ops=custom_mask_ops, verbose=verbose)
-
-
-    for m in hook_module_list:
-        m._buffers.pop("weight_mask")
-    # Remove registerd buffer on the model, and fixed following issue:
-    # https://github.com/Lyken17/pytorch-OpCounter/issues/96
-    for m in model.modules():
-        if 'total_ops' in m._buffers:
-            m._buffers.pop("total_ops")
-        if 'total_params' in m._buffers:
-            m._buffers.pop("total_params")
+    global register_ops, results
+    register_ops = mode_cls[mode]
+    if custom_ops is None:
+        custom_ops = {}
+    register_ops.update(custom_ops)
```

Comment:
What if I'm using different `custom_ops` for different models? Again, global is not safe.